### ISSUE TITLE
Intermittent unclean errors from twisted.test.test_ftp.FTPServerPasvDataConnectionTestCase.testTwoDirLIST on OS X

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2878]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2878|
|Type|defect|
|Created|2007-11-09 13:49:43Z|
|Branch|https://github.com/twisted/twisted/tree/ftp-test-cleanup-2878|

Seen occassionally:

```
===============================================================================
[ERROR]: twisted.test.test_ftp.FTPServerPasvDataConnectionTestCase.testTwoDirLIST

Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
DelayedCalls: (set twisted.internet.base.DelayedCall.debug = True to debug)
<DelayedCall 57837128 [599.546686888s] called=0 cancelled=0 FTP.__timedOut()>
===============================================================================
[ERROR]: twisted.test.test_ftp.FTPServerPasvDataConnectionTestCase.testTwoDirLIST

Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
Selectables:
<ProtocolWrapper #0 on 61530>
<<class 'twisted.internet.tcp.Port'> of twisted.protocols.ftp.DTPFactory on 61538>
-------------------------------------------------------------------------------
```

<details><summary>Searchable metadata</summary>

```
trac-id__2878 2878
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_ftp_test_cleanup_2878 branches/ftp-test-cleanup-2878
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1194616183000000 1194616183000000
changetime__1221250935000000 1221250935000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

A cleanup error occurs on OS X with another FTP test as well:

```
===============================================================================
[ERROR]: twisted.test.test_ftp.BasicFTPServerTestCase.testAnonymousLoginDenied

Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
DelayedCalls: (set twisted.internet.base.DelayedCall.debug = True to debug)
<DelayedCall 72203320 [599.86994195s] called=0 cancelled=0 FTP.__timedOut()>
===============================================================================
[ERROR]: twisted.test.test_ftp.BasicFTPServerTestCase.testAnonymousLoginDenied

Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
Selectables:
<ProtocolWrapper #0 on 57297>
-------------------------------------------------------------------------------
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [24790]) Branching to 'ftp-test-cleanup-2878'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

[#3405](https://github.com/twisted/twisted/issues/3405) was a duplicate of this, for the `testManyLargeDownloads` method.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner|
|-|-|

I think this is fixed.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @exarkun|
|-|-|

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [24811]) Merge ftp-test-cleanup-2878

Author: exarkun
Reviewer: therve
Fixes: [#2878](https://github.com/twisted/twisted/issues/2878)

Change the setUp of one of the base FTP test cases so that
the test will wait for the server-side of any FTP connection
established in the test to disconnect before ending the test,
avoiding leaving a selectable in the reactor for trial to
find and emit an unclean error for.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
